<a href="https://colab.research.google.com/github/adilsaleem007/Python-Basics-Assignment/blob/main/Assignment6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np

import re
import string

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 2. Load Dataset

In [9]:
df = pd.read_csv("/content/tweets.csv")
df.head()

,id,label,tweet
0,1,0,#fingerprint #Pregnancy Test https://goo.gl/h1...
1,2,0,Finally a transparant silicon case ^^ Thanks t...
2,3,0,We love this! Would you go? #talk #makememorie...
3,4,0,I'm wired I know I'm George I was made that wa...
4,5,1,What amazing service! Apple won't even talk to...


# 3. Data Preprocessing

In [10]:
def clean_text(text):
    text = text.lower()
    text = re.sub(r"http\S+", "", text)  # remove URLs
    text = re.sub(r"@\w+", "", text)     # remove mentions
    text = re.sub(r"#", "", text)        # remove hashtags
    text = re.sub(f"[{string.punctuation}]", "", text)
    return text

df['clean_text'] = df['tweet'].apply(clean_text)

### 4. Defining Features and Target Variable

In [12]:
X = df['clean_text']
y = df['label']

# 5. Train-Test Split

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 6. TF-IDF

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000)

X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

# 7. Train Model with Logistic Regression



In [22]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(class_weight='balanced')
model.fit(X_train_vec, y_train)

LogisticRegression(class_weight='balanced')

# 8. Predicting Model

In [23]:
y_pred = model.predict(X_test_vec)

# 9. Model Evaluation

In [24]:
from sklearn.metrics import accuracy_score, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))

Accuracy: 0.8800505050505051

Classification Report:

              precision    recall  f1-score   support

           0       0.96      0.87      0.91      1152
           1       0.72      0.91      0.80       432

    accuracy                           0.88      1584
   macro avg       0.84      0.89      0.86      1584
weighted avg       0.90      0.88      0.88      1584



# 10. Confusion Matrix

In [25]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_pred))

[[1003  149]
 [  41  391]]


# 11. Test Model

In [27]:
def predict_sentiment(text):
    text = clean_text(text)
    vec = vectorizer.transform([text])
    pred = model.predict(vec)
    return "Negative" if pred[0] == 1 else "Neutral/Positive" #some statements in the dataset are neutral

print(predict_sentiment("Great performance and smooth experience"))
print(predict_sentiment("Happy with this purchase"))
print(predict_sentiment("Worst battery ever"))

Neutral/Positive
Neutral/Positive
Neutral/Positive
Negative


# Insights
The model correctly classifies tweets into Negative and Neutral/Positive categories. It successfully identifies negative sentiments such as complaints, while treating general or positive tweets as Neutral/Positive.

# Conclusion
A sentiment analysis model was developed using TF-IDF and Logistic Regression. The model achieved an accuracy of approximately 88%, demonstrating good performance. The system is able to classify tweets into Negative and Neutral/Positive categories effectively.